quickwit-search/src/root.rs

```diff
@@ -209,13 +228,79 @@ pub async fn root_search(
     search_request_with_offset_0.max_hits += search_request.start_offset;
 
     // Perform the query phase.
-    let result_per_node_addr = execute_search(
+    let mut result_per_node_addr = execute_search(
         &assigned_leaf_search_jobs,
         search_request_with_offset_0.clone(),
     )
     .await?;
 
-    analyze_errors(&result_per_node_addr);
+    let analyze_result = analyze_errors(&result_per_node_addr);
+    if let Some(retry_action) = analyze_result.as_ref() {
+        // Create a job for fetching docs and assign the splits that the node is responsible for based on the job.
+        //
+        let leaf_search_jobs: Vec<Job> = job_for_split(
+            &retry_action.retry_split_ids.iter().collect(),
+            &split_metadata_map,
+        );
+
+        let retry_assigned_leaf_search_jobs = client_pool
+            .assign_jobs(leaf_search_jobs, &Some(retry_action.nodes_to_avoid.clone()))
+            .await?;
+        // Perform the query phase.
+        let result_per_node_addr_new = execute_search(
+            &retry_assigned_leaf_search_jobs,
+            search_request_with_offset_0.clone(),
+        )
+        .await?;
+
+        // clean out old errors
+        let complete_errors = result_per_node_addr
+            .iter()
+            .filter(|(_addr, res)| res.is_err())
+            .map(|(addr, _err)| *addr)
+            .collect_vec();
+        for err_addr in complete_errors {
+            result_per_node_addr.remove(&err_addr);
+        }
+        // remove partial, retryable errors
+        for result in result_per_node_addr.values_mut().flatten() {
+            let failed_splits: Vec<SplitSearchError> = std::mem::take(&mut vec![]);
+            result.failed_splits = failed_splits
+                .into_iter()
+                .filter(|err| !err.retryable_error)
+                .collect_vec();
+        }
+
+        for (addr, new_result) in result_per_node_addr_new {
+            if let Some(Ok(orig_res)) = result_per_node_addr.get_mut(&addr) {
```

Comment:
Was the borrow checker making `match` pattern impossible? (it often happens)

Reply:
I wasn't sure, what's more readable, but it's a match now